tests/metaworld/envs/mujoco/sawyer_xyz/base.py

```diff
@@ -0,0 +1,61 @@
+from metaworld.policies.base import PID
+
+
+def scripted_policy_test(env, policy, label, iters=1000):
+    """Tests whether a given policy solves an environment in a stateless manner
+
+    Args:
+        env: Environment to test
+        policy: Policy that's supposed to succeed in env
+        label: A string to use when printing out errors
+        iters: How many times the policy should be tested
+
+    """
+    for name, value in vars(policy).items():
+        assert isinstance(value, PID), label + ' policy has potential state variable: ' + name
+        assert value.stateful is False, label + ' policy has PID with stateful=True: ' + name
+
+    for i in range(iters):
+        ensure_success(env, policy)
+
+
+def ensure_success(env, policy, render=False, retries=1):
+    """Tests whether a given policy solves an environment
+
+    Args:
+        env: Environment to test
+        policy: Policy that's supposed to succeed in env
+        render: Whether to render the env in a GUI
+        retries: Number allowed attempts if max_path_length is exceeded
+
+    Returns:
+        int: Trajectory length
+    """
+    env.reset_model()
+    o = env.reset()
+
+    t = 0
+    done = False
+    success = False
+    while not success and not done:
+        a = policy.get_action(o)
+        try:
+            o, r, done, info = env.step(a)
+
+            if render:
+                env.render()
+
+            t += 1
+            success |= bool(info['success'])
+
+        # ValueError occurs when max_path_length is exceeded
+        # Re-trying helps avoid this 1 in 1000 failure edge case
+        except ValueError:
+            if retries > 0:
+                env.curr_path_length = 0
+                ensure_success(env, policy, render, retries-1)
```

Comment:
please do not write recursive tests. you can deal with this with an outer loop which report success rate (out of 100 trials, or n trials).

Reply:
Ok I can do that. I wrote it this way to try to avoid hard-coded thresholds (i.e. "We must pass 99/100") and instead test along the lines of, "We must not fail N times consecutively." But it does get a little messy so I'll change it.

Reply:
Hard-coded threshholds are exactly what tests are for :)